### PR TITLE
feat(shared-video): Allow only whitelisted URLs.

### PR DESCRIPTION
**Pull request by hristoterezov**

### CODE REVIEW

File: lang/main.json [hunk 1]
```diff
@@ -443,7 +443,7 @@
         "shareVideoTitle": "Share video",
         "shareYourScreen": "Share your screen",
         "shareYourScreenDisabled": "Screen sharing disabled.",
-        "sharedVideoDialogError": "Error: Invalid URL",
+        "sharedVideoDialogError": "Error: Invalid or forbidden URL",
```
Comment (saghul) [inline lang/main.json:446]: I'd undo this and add a log if you want, but keep things loose ;-)

File: react/features/shared-video/constants.ts [hunk 2]
```diff
@@ -28,3 +28,13 @@ export const PLAYBACK_STATUSES = {
     PAUSED: 'pause',
     STOPPED: 'stop'
 };
+
+/**
+ * The domain for youtube URLs.
+ */
+export const YOUTUBE_URL_DOMAIN = 'youtube.com';
```
Comment (saghul) [inline react/features/shared-video/constants.ts:35]: aren't there more domains like youtu.be and m.youtube.com and embed.yotube.com?

File: react/features/shared-video/middleware.any.ts [hunk 3]
```diff
@@ -62,9 +72,15 @@ MiddlewareRegistry.register(store => next => action => {
         break;
     }
     case CONFERENCE_LEFT:
+        if (!isSharedVideoEnabled(state)) {
```
Comment (saghul) [inline react/features/shared-video/middleware.any.ts:75]: Reverse the condition to put the dispatch inside and have a single break.

File: react/features/shared-video/functions.ts [hunk 4]
```diff
@@ -70,20 +76,71 @@ export function extractYoutubeIdOrURL(input: string) {
...
+ *
+ * @param {string} url - The URL.
+ * @returns {boolean}
+ */
+export function isURLAllowedForSharedVideo(url: string) {
+    if (!url) {
+        return false;
+    }
+
+    if (!url.match(/http/)) {
```
Comment (saghul) [inline react/features/shared-video/functions.ts:135]: You want to match from the start `^` but allso what about casing? `hTTpS` ? Isn't is better to just parse the URL all the time and act on the exception?

File: react/features/shared-video/functions.ts [hunk 5]
```diff
@@ -70,20 +76,71 @@ export function extractYoutubeIdOrURL(input: string) {
...
+        }
     }
 
     // Check if the URL is valid, native may crash otherwise.
     try {
         // eslint-disable-next-line no-new
-        new URL(trimmedLink);
+        const url = new URL(trimmedLink);
+
+        if (!url?.hostname || !URL_WHITELIST.includes(url?.hostname)) {
```
Comment (saghul) [inline react/features/shared-video/functions.ts:92]: Isn't the first check redundant?

File: react/features/shared-video/constants.ts (hunk 2 above)
Comment (hristoterezov) [reply react/features/shared-video/constants.ts:35]: There are! 

I use this as a placeholder and don't use it to match on it. Basically checking if this string is there then youtube is enabled, if not then it is disabled.

File: react/features/shared-video/functions.ts (hunk 5 above)
Comment (hristoterezov) [reply react/features/shared-video/functions.ts:92]: not really:

image

Comment (hristoterezov) [reply react/features/shared-video/functions.ts:92]: Actually you are correct.

**Review by saghul — APPROVED** [review]

**Pull request merged**